Terraform major upgrade

seriva opened the pull request:
- Bumped Terraform from 0.12.6 to 1.1.3.
- Bumped azurerm provider from 1.38 to 2.91.
- Bumped AWS provider from 2.26 to 3.70.1
- Updated azurerm provider scripts
- Updated AWS provider scripts
- Updated Azure-cli from 2.29 to 2.32
- Fix for #1569.
- Added Terraform manual update documentation
- Added incompatible terraform check
- Added changelog.
- Bumped version to 2.0.0dev
- Apply terraform fmt to enforce standard on templates

seriva committed:
Upgraded terraform and azure-cli.

seriva committed:
Fix terraform command.

seriva committed:
Minor fix to terraform command.

seriva committed:
Fix Azure terraform to work with the latest

seriva committed:
Fix Azure provider unittests

seriva committed:
Fix for terraform azure resource destruction.

seriva committed:
Increased ping-retries to 10.

seriva committed:
Pinned azurerm to latest.

seriva committed:
Pinned AWS provider.

seriva committed:
Added terraform version arg to release container

seriva committed:
Bump terraform to the latest.

seriva committed:
Bumped version nr.

seriva committed:
Added Terraform upgrade documentation for Azure

seriva committed:
Added Terraform upgrade documentation for AWS

seriva committed:
Added changelog

seriva committed:
Fixed #1569

seriva committed:
Added changelog link

seriva committed:
Minor fix for #1569

seriva committed:
Added incompatible terraform check.

seriva committed:
Fixed space.

seriva committed:
Fixed typo.

seriva committed:
Minor document addition.

atsikham commented on docs/home/howto/UPGRADE.md:
```diff
@@ -457,3 +457,166 @@ directories by running `delete_old_cluster.sh` script (created in `pg_upgrade_wo
+- Azurerm provider 1.38.0 to 2.91.0
+- AWS provider 2.26 to 3.71.0
+
+These introduce some breaking changes which will require manual steps for upgrading an existing 1.x clusters. As this is not straight forward we recommand deploying a new cluster on 2.x and migrating data instead.
```
```suggestion
These introduce some breaking changes which will require manual steps for upgrading an existing 1.x clusters. As this is not straight forward we recommend deploying a new cluster on 2.x and migrating data instead.
```

atsikham commented on terraform/azure/infrastructure/network-interface-security-group-association.j2:
```diff
@@ -0,0 +1,16 @@
+resource "azurerm_network_interface_security_group_association" "{{ specification.name }}" {
+  network_interface_id      = "${azurerm_network_interface.{{ specification.network_interface_name }}.id}"
+  network_security_group_id = "${azurerm_network_security_group.{{ specification.security_group_name }}.id}"
+}
```
newline

atsikham reviewed (changes requested):
In addition to minor changes suggested: did you use [terraform fmt](https://www.terraform.io/cli/commands/fmt)? If no, it might be the right PR to do. Continue to check.

cicharka commented on docs/home/howto/UPGRADE.md:
```diff
@@ -457,3 +457,166 @@ directories by running `delete_old_cluster.sh` script (created in `pg_upgrade_wo
+
+#### v0.12.6 => v0.13.x
+
+Offical documentation can be found here: https://www.terraform.io/language/upgrade-guides/0-13
```
```suggestion
Official documentation can be found here: https://www.terraform.io/language/upgrade-guides/0-13
```

cicharka replied:
And it occurs in more lines below.

cicharka replied:
There are occurrences of `Offical` still in lines: 461. 476, 491, 527, 541, 556, 570 in the same file. These lines should be corrected as well.

seriva replied:
All fixed.

erzetpe commented on docs/home/howto/UPGRADE.md:
```diff
@@ -457,3 +457,166 @@ directories by running `delete_old_cluster.sh` script (created in `pg_upgrade_wo
+
+These introduce some breaking changes which will require manual steps for upgrading an existing 1.x clusters. As this is not straight forward we recommand deploying a new cluster on 2.x and migrating data instead.
+
+If you do want to upgrade a 1.x manually it will require you to upgrade between a few different versions of Terraform. The basic steps are described below for each provider. As always insure backup of any data required.
```
```suggestion
If you do want to upgrade a 1.x manually it will require you to upgrade between a few different versions of Terraform. The basic steps are described below for each provider. As always ensure backup of any data required.
```

erzetpe reviewed (changes requested):
Additionally perform rebase from develop please.

seriva committed:
Update docs/home/howto/UPGRADE.md

Co-authored-by: Anatoli Tsikhamirau <anatoliytihomirov@yahoo.com>

seriva committed:
Update docs/home/howto/UPGRADE.md

Co-authored-by: erzetpe <2865776+erzetpe@users.noreply.github.com>

seriva committed:
Update docs/home/howto/UPGRADE.md

Co-authored-by: cicharka <93913624+cicharka@users.noreply.github.com>

seriva committed:
Merge branch 'develop' into feature/2825

seriva committed:
Fix newline

seriva committed:
Fixed documentation.

seriva commented:
> In addition to minor changes suggested: did you use [terraform fmt](https://www.terraform.io/cli/commands/fmt)? If no, it might be the right PR to do. Continue to check.

Ill check it.

erzetpe commented on docs/changelogs/CHANGELOG-2.0.md:
```diff
@@ -17,6 +24,9 @@
+
 ### Breaking changes
 
+- Upgrade of Terraform components in issue [#2825](https://github.com/epiphany-platform/epiphany/issues/2825) will make running re-apply with infrastructure break on existing 1.x clusters. Advice is deploy a new cluster and migrate data. If needed a manual upgrade path is described [here.](../home/howto/UPGRADE.md#terraform-upgrade-from-epiphany-1.x-to-2.x)
```
```suggestion
- Upgrade of Terraform components in issue [#2825](https://github.com/epiphany-platform/epiphany/issues/2825) will make running re-apply with infrastructure break on existing 1.x clusters. The advice is to deploy a new cluster and migrate data. If needed a manual upgrade path is described [here.](../home/howto/UPGRADE.md#terraform-upgrade-from-epiphany-1.x-to-2.x)
```

erzetpe commented on docs/changelogs/CHANGELOG-2.0.md:
```diff
@@ -1,13 +1,20 @@
+
 ### Updated
 
+- [#2825](https://github.com/epiphany-platform/epiphany/issues/2825) - Research upgrade for Terraform
```
Shouldn't be updated to refer actual aim of this task?

erzetpe commented on cli/engine/ApplyEngine.py:
```diff
@@ -99,6 +97,14 @@ def load_manifest(self):
         if os.path.isfile(path_to_manifest):
             self.manifest_docs = load_manifest(get_build_path(self.cluster_model.specification.name))
 
+    def assert_incompatible_terraform(self):
```
Maybe this method name can be changed to something like check_if_incompatible_terraform_version?

seriva replied:
We want to assert not just check so leaving as is.

erzetpe commented on docs/home/howto/UPGRADE.md:
```diff
@@ -421,3 +421,166 @@ directories by running `delete_old_cluster.sh` script (created in `pg_upgrade_wo
+
+#### v0.12.6 => v0.13.x
+
+Official documentation can be found here: https://www.terraform.io/language/upgrade-guides/0-13
```
```suggestion
The official documentation can be found here: https://www.terraform.io/language/upgrade-guides/0-13
```

erzetpe commented on docs/home/howto/UPGRADE.md:
```diff
@@ -421,3 +421,166 @@ directories by running `delete_old_cluster.sh` script (created in `pg_upgrade_wo
+
+Note: From v0.14.x we can upgrade straight to v1.0.x. No need to upgrade to v0.15.x first.
+
+Official documentation can be found here: https://www.terraform.io/language/upgrade-guides/1-0
```
```suggestion
The official documentation can be found here: https://www.terraform.io/language/upgrade-guides/1-0
```

erzetpe commented on docs/home/howto/UPGRADE.md:
```diff
@@ -421,3 +421,166 @@ directories by running `delete_old_cluster.sh` script (created in `pg_upgrade_wo
+
+In this step we also force the upgrade from Azurerm provider 1.38.0 to 2.91.0 which requires a few more steps to resolve some pending issues.
+
+Official documentation can be found here: https://www.terraform.io/language/upgrade-guides/1-1
```
```suggestion
The official documentation can be found here: https://www.terraform.io/language/upgrade-guides/1-1
```

erzetpe commented on docs/home/howto/UPGRADE.md:
```diff
@@ -421,3 +421,166 @@ directories by running `delete_old_cluster.sh` script (created in `pg_upgrade_wo
+
+#### v0.12.6 => v0.13.x
+
+Official documentation can be found here: https://www.terraform.io/language/upgrade-guides/0-13
```
```suggestion
The official documentation can be found here: https://www.terraform.io/language/upgrade-guides/0-13
```

erzetpe commented on docs/home/howto/UPGRADE.md:
```diff
@@ -421,3 +421,166 @@ directories by running `delete_old_cluster.sh` script (created in `pg_upgrade_wo
+
+#### v0.13.x => v0.14.x
+
+Official documentation can be found here: https://www.terraform.io/language/upgrade-guides/0-14
```
```suggestion
The official documentation can be found here: https://www.terraform.io/language/upgrade-guides/0-14
```

erzetpe commented on docs/home/howto/UPGRADE.md:
```diff
@@ -421,3 +421,166 @@ directories by running `delete_old_cluster.sh` script (created in `pg_upgrade_wo
+
+Note: From v0.14.x we can upgrade straight to v1.0.x. No need to upgrade to v0.15.x first.
+
+Official documentation can be found here: https://www.terraform.io/language/upgrade-guides/1-0
```
```suggestion
The official documentation can be found here: https://www.terraform.io/language/upgrade-guides/1-0
```

erzetpe commented on docs/home/howto/UPGRADE.md:
```diff
@@ -421,3 +421,166 @@ directories by running `delete_old_cluster.sh` script (created in `pg_upgrade_wo
+
+In this step we also force the upgrade from AWS provider 2.26 to 3.71.0 which requires a few more steps to resolve some pending issues.
+
+Official documentation can be found here: https://www.terraform.io/language/upgrade-guides/1-1
```
```suggestion
The official documentation can be found here: https://www.terraform.io/language/upgrade-guides/1-1
```

seriva committed:
Fix documentation.

seriva committed:
Update docs/changelogs/CHANGELOG-2.0.md

Co-authored-by: erzetpe <2865776+erzetpe@users.noreply.github.com>

seriva committed:
Fixes for changelog.

seriva committed:
Use terraform fmt to fix a few issues on Azure.

seriva committed:
Merge branch 'develop' into feature/2825

seriva commented:
> Additionally perform rebase from develop please.

Done.

erzetpe commented on docs/changelogs/CHANGELOG-2.0.md:
```diff
@@ -1,13 +1,22 @@
+
 ### Fixed
 
+[#1569](https://github.com/epiphany-platform/epiphany/issues/1569) - Azure unmanaged disks not suppored by Epiphany but there is misleading setting in the default configuration
```
```suggestion
[#1569](https://github.com/epiphany-platform/epiphany/issues/1569) - Azure unmanaged disks not supported by Epiphany but there is misleading setting in the default configuration
```

seriva committed:
Apply terraform fmt for AWS provider.

seriva committed:
Update docs/changelogs/CHANGELOG-2.0.md

Co-authored-by: erzetpe <2865776+erzetpe@users.noreply.github.com>

seriva commented:
> > In addition to minor changes suggested: did you use [terraform fmt](https://www.terraform.io/cli/commands/fmt)? If no, it might be the right PR to do. Continue to check.
> 
> Ill check it.

Applied to both AWS and Azure templates.

cicharka reviewed (dismissed):
LGTM

cicharka commented:
/azp run

seriva committed:
Minor import fix.

cicharka committed:
* UPGRADE.md docs minor changes
* query_yes_no for assert_incompatible_terraform added

cicharka committed:
Remove commented TODO

This template when paresd by TerraformTemplateGenerator is failing due
to: `ERROR epicli - 'cli.helpers.ObjDict.ObjDict object' has no
attribute 'storage_container'`. Removal of commented lines helps parser
to fulfil its job.

cicharka commented:
/azp run

atsikham reviewed (approved):
LGTM

cicharka commented:
Tested with:
  provider: Azure:
  ✔️ apply v2.0 (all components and apps enabled)
  ✔️ apply v1.3, manual upgrade of Terraform, re-apply v2.0 (all components and apps enabled/minimal cluster)
  ✔️ delete v2.0
  provider: AWS:
  ✔️ apply v2.0 (all components and apps enabled)
  ✔️ delete v2.0

cicharka commented:
Tested:
- provider **Azure**:
	- **Ubuntu** and **RedHat** (all components and apps enabled):
		- init v2.0dev -> apply v2.0dev -> delete v2.0dev
		- upgrade from v1.0.1 -> v2.0dev:
			- includes manual Terraform upgrade
			- re-applying and re-upgrade to test idempotency
			- backup and recovery commands
			- delete
- provider **AWS**:
	- **Ubuntu** and **RedHat** (all components and apps enabled):
		- init v2.0dev -> apply v2.0dev -> delete v2.0de
All scenarios verified with `epicli test` .
